Failing lightning-cli tests

### Bug description

CI on the main branch fails with this error:

```
_____________ test_lightning_cli[_trainer_builder-_model_builder] _____________

trainer_class = <function _trainer_builder at 0x000002310D3008B0>
model_class = <function _model_builder at 0x000002310D300790>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x0000023115A25AB0>

    @pytest.mark.parametrize(("trainer_class", "model_class"), [(Trainer, Model), (_trainer_builder, _model_builder)])
    def test_lightning_cli(trainer_class, model_class, monkeypatch):
        """Test that LightningCLI correctly instantiates model, trainer and calls fit."""
        expected_model = {"model_param": 7}
        expected_trainer = {"limit_train_batches": 100}
    
        def fit(trainer, model):
            for k, v in expected_model.items():
                assert getattr(model, k) == v
            for k, v in expected_trainer.items():
                assert getattr(trainer, k) == v
            save_callback = [x for x in trainer.callbacks if isinstance(x, SaveConfigCallback)]
            assert len(save_callback) == 1
            save_callback[0].on_train_start(trainer, model)
    
        def on_train_start(callback, trainer, _):
            config_dump = callback.parser.dump(callback.config, skip_none=False)
            for k, v in expected_model.items():
                assert f"  {k}: {v}" in config_dump
            for k, v in expected_trainer.items():
                assert f"  {k}: {v}" in config_dump
            trainer.ran_asserts = True
    
        monkeypatch.setattr(Trainer, "fit", fit)
        monkeypatch.setattr(SaveConfigCallback, "on_train_start", on_train_start)
    
        with mock.patch("sys.argv", ["any.py", "fit", "--model.model_param=7", "--trainer.limit_train_batches=100"]):
>           cli = LightningCLI(model_class, trainer_class=trainer_class, save_config_callback=SaveConfigCallback)

test_cli.py:153: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:368: in __init__
    self.setup_parser(run, main_kwargs, subparser_kwargs)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:403: in setup_parser
    self._add_subcommands(self.parser, **subparser_kwargs)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:480: in _add_subcommands
    subcommand_parser = self._prepare_subcommand_parser(trainer_class, subcommand, **subparser_kwargs)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:486: in _prepare_subcommand_parser
    self._add_arguments(parser)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:438: in _add_arguments
    self.add_core_arguments_to_parser(parser)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:421: in add_core_arguments_to_parser
    parser.add_lightning_class_args(self.trainer_class, "trainer")
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\lightning\pytorch\cli.py:125: in add_lightning_class_args
    lightning_class = class_from_function(lightning_class)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

func = <function _trainer_builder at 0x000002310D3008B0>
func_return = <class 'lightning.pytorch.trainer.trainer.Trainer'>
name = '_trainer_builder_class'

    def class_from_function(
        func: Callable[..., ClassType],
        func_return: Optional[Type[ClassType]] = None,
        name: Optional[str] = None,
    ) -> Type[ClassType]:
        """Creates a dynamic class which if instantiated is equivalent to calling func.
    
        Args:
            func: A function that returns an instance of a class.
            func_return: The return type of the function. Required if func does not have a return type annotation.
            name: The name of the class. Defaults to function name suffixed with "_class".
        """
        if func_return is None:
            func_return = inspect.signature(func).return_annotation
        if func_return is inspect.Signature.empty:
            raise ValueError(f"{func} does not have a return type annotation")
        if isinstance(func_return, str):
            try:
                func_return = get_type_hints(func)["return"]
                if sys.version_info[:2] != (3, 6) and isinstance(func_return, __import__("typing").ForwardRef):
                    func_return = func_return._evaluate(func.__globals__, {})
            except Exception as ex:
                func_return = inspect.signature(func).return_annotation
                raise ValueError(f"Unable to dereference {func_return}, the return type of {func}: {ex}") from ex
    
        if not name:
            name = func.__qualname__.replace(".", "__") + "_class"
    
        caller_module = inspect.getmodule(inspect.stack()[1][0]) or inspect.getmodule(class_from_function)
        assert caller_module
        if hasattr(caller_module, name):
>           raise ValueError(f"{caller_module.__name__} already defines {name!r}, please use a different name")
E           ValueError: lightning.pytorch.cli already defines '_trainer_builder_class', please use a different name

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\jsonargparse\_util.py:397: ValueError
```
Full job:
https://github.com/Lightning-AI/lightning/actions/runs/5680049817/job/15393384681


For some strange reason, the error does not occur on PRs at the moment? We need to investigate this

### What version are you seeing the problem on?

master



cc @carmocca @borda @mauvilsa

--- a/requirements/pytorch/extra.txt
+++ b/requirements/pytorch/extra.txt
@@ -5,6 +5,6 @@
 matplotlib>3.1, <3.6.2
 omegaconf >=2.0.5, <2.4.0
 hydra-core >=1.0.5, <1.4.0
-jsonargparse[signatures] >=4.18.0, <4.23.0
+jsonargparse[signatures] >=4.18.0, <4.23.0  # strict
 rich >=12.3.0, <=13.4.2
 tensorboardX >=2.2, <=2.6.1  # min version is set by torch.onnx missing attribute

